### PR TITLE
aptos: init at unstable-2022-08-26

### PR DESCRIPTION
**Pull request by macalinao**
###### Description of changes

Adds Aptos package

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/blockchains/aptos/default.nix [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        inherit (llvmPackages_12) llvm clang libclang;
+      };
+      rustPlatform.buildRustPackage rec {
+        inherit pname cargoBuildFlags version cargoSha256;
+        verifyCargoDeps = true;
+
+        src = fetchFromGitHub {
+          inherit rev sha256;
+          owner = "aptos-labs";
+          repo = "aptos-core";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:39]:
```suggestion
          owner = "aptos-labs";
          repo = "aptos-core";
          # https://github.com/aptos-labs/aptos-core/tree/devnet
          rev = "8399cd1c7b9662d3a6a09c28363c5f66f0839c41";
          sha256 = "sha256-XBBX58V+Z+p5ajLwCMlpFGLwe+HEBmyLwqIpDcIq8xU=";
```

File: pkgs/applications/blockchains/aptos/default.nix [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+, libiconv
+, symlinkJoin
+}:
+
+let
+  version = "unstable-2022-08-26";
+  # https://github.com/aptos-labs/aptos-core/tree/devnet
+  rev = "8399cd1c7b9662d3a6a09c28363c5f66f0839c41";
+  sha256 = "sha256-XBBX58V+Z+p5ajLwCMlpFGLwe+HEBmyLwqIpDcIq8xU=";
+
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:23]:
```suggestion
```

File: pkgs/applications/blockchains/aptos/default.nix [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  buildAptos =
+    { pname
+    , cargoSha256
+    , cargoBuildFlags ? [ ]
+    }:
+      with {
+        inherit (llvmPackages_12) llvm clang libclang;
+      };
+      rustPlatform.buildRustPackage rec {
+        inherit pname cargoBuildFlags version cargoSha256;
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:33]:
```suggestion
        inherit pname cargoBuildFlags cargoSha256;
        version = "unstable-2022-08-26";
```

File: pkgs/applications/blockchains/aptos/default.nix [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        # Skip tests for now
+        doCheck = false;
+
+        # Needed to get openssl-sys to use pkg-config.
+        OPENSSL_NO_VENDOR = 1;
+        OPENSSL_LIB_DIR = "${lib.getLib openssl}/lib";
+        OPENSSL_DIR = "${lib.getDev openssl}";
+
+        # ensure we are using LLVM to compile
+        LLVM_CONFIG_PATH = "${llvm}/bin/llvm-config";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:54]:
```suggestion
        LLVM_CONFIG_PATH = "${llvmPackages_12.llvm}/bin/llvm-config";
```

File: pkgs/applications/blockchains/aptos/default.nix [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+        # see https://github.com/NixOS/nixpkgs/issues/52447#issuecomment-852079285
+        LIBCLANG_PATH = "${libclang.lib}/lib";
+        BINDGEN_EXTRA_CLANG_ARGS =
+          "-isystem ${libclang.lib}/lib/clang/${lib.getVersion clang}/include";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:59]:
```suggestion
        LIBCLANG_PATH = "${llvmPackages_12.libclang.lib}/lib";
        BINDGEN_EXTRA_CLANG_ARGS = with llvmPackages_12; "-isystem ${libclang.lib}/lib/clang/${lib.getVersion clang}/include";
```

File: pkgs/applications/blockchains/aptos/default.nix [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        ROCKSDB_LIB_DIR = "${rocksdb}/lib";
+
+        nativeBuildInputs = [
+          pkg-config
+          rustc
+          cargo
+          clang
+
+          rustfmt
+          llvm
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:73]:
```suggestion
          llvmPackages_12.clang
          llvmPackages_12.llvm
```

File: pkgs/applications/blockchains/aptos/default.nix [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        verifyCargoDeps = true;
+
+        src = fetchFromGitHub {
+          inherit rev sha256;
+          owner = "aptos-labs";
+          repo = "aptos-core";
+        };
+
+        PKG_CONFIG_PATH = "${openssl.dev}/lib/pkgconfig";
+        RUST_SRC_PATH = "${rustPlatform.rustLibSrc}";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:43]:
```suggestion
        PKG_CONFIG_PATH = "${lib.getDev openssl}/lib/pkgconfig";
        RUST_SRC_PATH = rustPlatform.rustLibSrc;
```

File: pkgs/applications/blockchains/aptos/default.nix [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+        # Skip tests for now
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:45]: but why? Do they not easily work?

File: pkgs/applications/blockchains/aptos/default.nix [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        # see: https://github.com/aptos-labs/aptos-core/blob/36dfc6499dd576d7d2ba883b66161510ff5cbe6b/.circleci/config.yml#L241
+        buildInputs = [
+          rocksdb
+          postgresql # libpq
+          openssl # libssl
+        ] ++ (
+          lib.optional stdenv.isDarwin ([ libiconv ]
+            ++ (with darwin.apple_sdk.frameworks; [ DiskArbitration Foundation ]))
+        );
+        strictDeps = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:85]:
```suggestion
```
Already done here https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/build-rust-package/default.nix#L151

File: pkgs/applications/blockchains/aptos/default.nix [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+        # see: https://github.com/aptos-labs/aptos-core/blob/36dfc6499dd576d7d2ba883b66161510ff5cbe6b/.circleci/config.yml#L241
+        buildInputs = [
+          rocksdb
+          postgresql # libpq
+          openssl # libssl
+        ] ++ (
+          lib.optional stdenv.isDarwin ([ libiconv ]
+            ++ (with darwin.apple_sdk.frameworks; [ DiskArbitration Foundation ]))
+        );
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:84]:
```suggestion
        ] ++ lib.optional stdenv.isDarwin [
          libiconv DiskArbitration Foundation
        ];
```
Please add those to inputs and inherit them in top-level.

**Comment by SuperSandro2000** [comment]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

**Comment by macalinao** [comment]
Updated

File: pkgs/applications/blockchains/aptos/default.nix [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        rocksdb
+        postgresql # libpq
+        openssl # libssl
+      ] ++ (
+        lib.optional stdenv.isDarwin [
+          libiconv
+          DiskArbitration
+          Foundation
+        ]
+      );
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:87]:
```suggestion
      ] ++ lib.optional stdenv.isDarwin [
        libiconv
        DiskArbitration
        Foundation
      ];
```

Comment (macalinao) [reply pkgs/applications/blockchains/aptos/default.nix]: Updated

File: pkgs/applications/blockchains/aptos/default.nix [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+, Foundation
+}:
+
+let
+  cargoHashes = {
+    aptos = "sha256-y9wCwMCtvCu4MzW18MXZ2V+wwhuMkHF2hcevdvW17k4=";
+    aptos-node = "sha256-gQabtMi7FpVCJJYeTLebL4d4fQMO6qf9Hhgj1vltR90=";
+    aptos-tools = "sha256-FtSOkhHghhnIrrInpGi8n9rWjOB73fjHEBs3lpWbr20=";
+  };
+
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/aptos/default.nix:25]: Why are we not just inlining this?

**Pull request closed without merging**